src/libs/auth.js

```diff
@@ -411,32 +419,57 @@ authStore.subscribe(withErrorReporting('Error loading Framework Services account
   }
 }))
 
+export const updateFenceLinkExpirationNotification = (provider, status) => {
+  const { key, name } = provider
+
+  // Clear any old notifications
+  clearMatchingNotifications(`fence-link-expiration/${key}/`)
+
+  const now = Date.now()
+  // Bond API returns link time as an ISO formatted string.
+  const dateOfExpiration = status && addDays(provider.expiresAfter, parseJSON(status.issued_at))
+  const shouldNotify = Boolean(dateOfExpiration) && now >= addDays(-5, dateOfExpiration)
+
+  if (shouldNotify) {
+    const hasExpired = now >= dateOfExpiration
+
+    // Separate notifications for expired and expiring. If a user mutes the notification that a link will
+    // expire soon, then we still want to notify them once the link has actually expired.
+    // The link expiration time is included in the notification ID so that a muting notifications only
+    // applies to the current instance of the link. If an account is re-linked, notifications should
+    // be shown again when that link is expiring.
+    const notificationId = `fence-link-expiration/${key}/${dateOfExpiration.getTime()}/${hasExpired ? 'expired' : 'expiring'}`
```

Comment:
Consider extracting this line into a function, which will provide a single location for the comment.

Reply:
I tried this and found it kind of awkward. The notification IDs generated here need to match the prefix passed to `clearMatchingNotifications`, so it felt like those prefixes should also be extracted. And I'd either have to pass both `dateOfExpiration` and `hasExpired` (which is derived from `dateOfExpiration`) to the function or recalculate `hasExpired`.

Reply:
Hmm... I think you could manage the key issue, but yeah about the others:
```javascript
const getNotificationIdPath = (dateOfExpiration, hasExpired) =>
  // comment about expired vs. expiring
  `${dateOfExpiration.getTime()}/${hasExpired ? 'expired' : 'expiring'}`
```
Usage:
```javascript
const notificationId = `fence-link-expiration/${key}/${getNotificationIdPath(dateOfExpiration, hasExpired)}`
```
I agree. Not an obvious win.